scipy/sparse/construct.py

```diff
@@ -664,16 +664,27 @@ def block_diag(mats, format=None, dtype=None):
            [0, 0, 0, 7]])
 
     """
-    nmat = len(mats)
-    rows = []
+    dtype = np.dtype(dtype)
+    row = []
+    col = []
+    data = []
+    r_idx = 0
+    c_idx = 0
     for ia, a in enumerate(mats):
-        row = [None]*nmat
         if issparse(a):
-            row[ia] = a
+            a = a.tocsr()
         else:
-            row[ia] = coo_matrix(a)
-        rows.append(row)
-    return bmat(rows, format=format, dtype=dtype)
+            a = coo_matrix(a).tocsr()
+        nrows, ncols = a.shape
+        for r in range(nrows):
```

Comment:
Using `a.nonzero()` seems like an easier and faster approach here.

Reply:
I just tried this and it gives an incorrect result when the last columns or rows of the last matrix are zeros. It will omit these rows/columns, resulting in a smaller matrix than expected. The simplest case is included in the unit tests, where the input and expected output is [[1, 0]]. If I use the non-zeros, I get [[1]].

In any case, I think the for loop can be improved by removing the if condition, which is actually not necessary:
```python
        for r in range(nrows):
            for c in range(ncols):
                row.append(r + r_idx)
                col.append(c + c_idx)
                data.append(a[r, c])
```

Reply:
This goes over every entry of the sparse matrix. Instead what @perimosocordiae suggests is to go over only the data of the matrix. Something along the lines of the following should be working

    for r, c in zip(*a.nonzero()):
        data.append(a[r, c])
        ....

Reply:
Yes, that's what I tried. The problem is that if the input of block_diag is [[1, 0]], then the 0 is not appended, so the output becomes [[1]]. One of the unit tests tests for this.

Reply:
Ahh, I did not realize I could pass the shape to coo_matrix when building the full matrix. (I am not a regular user of scipy, so I am not that familiar with the API. Please bear with me :sweat_smile:)